### ISSUE TITLE
All tabs in STG are lost if icon is clicked prior to loading sequence

### ISSUE THREAD

**Issue by zkhcohen** [Drive4ik/simple-tab-groups#1035]
**Describe the bug**
If you start Firefox and click the STG icon in the extensions bar before the initial loading sequence (black, rotating loading icon), all of the tabs in the STGs will be lost. The STGs themselves will remain, but the tab count will appear as "0" in each STG. Clicking on the tab groups or restarting Firefox does not restore the tabs. The only solution is to restore from backup.

**To Reproduce**
1. Open Firefox. Wait a moment for STG to load.
2. Close Firefox.
3. Quickly open Firefox and click on the STG icon as fast as possible. If done successfully, when you click on STG again, all of your tab groups will be gone.

**Expected behavior**
Following the steps above would not result in the loss of your saved tabs.

**Desktop (please complete the following information):**
 - OS and version: Windows 11
 - Firefox version: 112.0.1 
 - Simple Tab Groups version: 5.1

**Additional context**
I'll try to do some additional testing to narrow down the root cause. Let me know if logs or settings would assist in the troubleshooting process.

**Comment by Drive4ik** [comment]
Yes, you can enable debug mode in the settings and try to reproduce the problem. You can email me the log with a link to this ticket. I can't reproduce the problem yet.
You can even record a GIF animation)) https://www.cockos.com/licecap/

**Comment by zkhcohen** [comment]
> Yes, you can enable debug mode in the settings and try to reproduce the problem. You can email me the log with a link to this ticket. I can't reproduce the problem yet. You can even record a GIF animation)) https://www.cockos.com/licecap/

Sent logs.

[Gif here](https://www.mediafire.com/view/kfaxumk7fenimcm)

**Comment by rifat0153** [comment]
I am facing a similar issue. After opening a tab group, while it's loading, if I close Firefox during loading I will lose all my saved tabs. 

[https://firebasestorage.googleapis.com/v0/b/portfolio-6b79f.appspot.com/o/Animation.gif?alt=media&token=5c2faaf0-1216-4ae6-a1be-d958c7a2ee8f](Download Gif Here)

Firefox: 112.0.2 (64-bit)
STG: 5.2
OS: Windows 11

**Comment by zkhcohen** [comment]
Some additional notes:
1. This also occurs when starting Firefox in Private Browsing Mode and then switching back to a normal session. In Private Browsing, a notification pops up saying that Simple Tab Groups couldn't be started. This extension isn't enabled in Private Browsing to begin with.
2. The 'lost tabs' issue goes back as far as Version 4.7.2.1 from my testing, but likely dates even farther back, potentially since its inception.
3. The core issue appears to be around the retrieval of tabs to restore in stg-background.js. In particular, the following function fails on the final return in the following code snippet:

/addon/src/stg-background.js:
``` javascript
let restoringMissedTabsPromise = null; // need when remove window
async function tryRestoreMissedTabs() {
    const log = logger.start(['info', 'tryRestoreMissedTabs']);

    if (restoringMissedTabsPromise) {
        await restoringMissedTabsPromise.catch(() => { });
    }

    let [
        { tabsToRestore: tabsToRestoreNotModified },
        windows,
    ] = await Promise.all([
        Storage.get({ tabsToRestore: [] }),
        Windows.load(),
    ]);

    if (!tabsToRestoreNotModified.length || !windows.length) {
        log.stop('not found tabs/windows for restore');
        return;
    }
```

I'll continue debugging as time allows. Hopefully I can find a fix that can be implemented in a fork I created.

**Comment by zkhcohen** [comment]
Update 5/12/2024: The issue occurs without any other extensions added to Firefox, and without any of the default Firefox settings changed. I've also found that a reliable way to replicate the issue is to start and then close Firefox over and over again until the tabs disappear from the groups.

I think I'm getting closer to tracking down the specific culprit in the handling of storage. I also have a working dev build that I'm modifying, deploying and debugging locally. Note for anyone else who tries to build this extension locally, you need NPM 17.9.1 - ver. 22.1.0 doesn't work.

I think the stack looks something like this (with my own presumptive comments added):

At the top-level, the highlighted line in the following snippet returns nothing, causing the tab restoration process to abort:
``` javascript
    let [
        { tabsToRestore: tabsToRestoreNotModified },
        windows,
    ] = await Promise.all([
         // set tabsToRestoreNotModified equal to tabsToRestore
>       Storage.get({ tabsToRestore: [] }), 
        Windows.load(),
    ]);
```

Another level deeper, the highlighted line removes the "tabsToRestore" key from the local storage database. Commenting this out fixes the issue. Since this either requires "tabsToRestoreNotModified" == "tabsToRestore", and "tabsToRestoreNotModified" is defined by "tabsToRestore" above, prior to the restoration process, this either means that "tabsToRestore" still equals "tabsToRestoreNotModified", or, based on the logic below, "tabsToRestore" was emptied:
``` javascript
// set tabsInDB equal to tabsToRestore from local storage. we know tabsToRestore can't be null, otherwise we wouldn't have gotten this far into the function.
let { tabsToRestore: tabsInDB } = await Storage.get({ tabsToRestore: [] });
    // filter tabsInDB based on values in tabsToRestoreNotModified
    tabsInDB = tabsInDB.filter(t => {
        // filter out all of the tabs that exist in both lists, leaving just the unique tabs
        return !tabsToRestoreNotModified.some(tab => t.groupId === tab.groupId && t.url === tab.url && t.cookieStoreId === tab.cookieStoreId);
    });

    // if the lists are the same or tabsInDB is an empty array (due to tabsToRestore being empty), the length will be 0
    if (tabsInDB.length) {
        // if there are tabs that haven't been restored, set tabsToRestore to those tabs
        await Storage.set({ tabsToRestore: tabsInDB });
    } else {
           // if there are no tabs left to restore (because the lists match), then remove the tabsToRestore from the local storage DB
>        await Storage.remove('tabsToRestore');
    }
```

The lack of documented code is making it extremely difficult to determine exactly what the root cause is, but I'm fairly confident it was introduced with the following commit, at least symptomatically: https://github.com/Drive4ik/simple-tab-groups/commit/0f9e1850480ed7011ad88dee87be8a0419b34fc5

Cross-referencing it with the logs, it appears that even on runs where ```Storage.remove('tabsToRestore');``` was called, subsequent runs don't necessarily fail. Maybe this suggests that the mechanism which is supposed to re-set the tabsToRestore fails in certain cases, and by _not_ removing tabsToRestore from the database, it serves as a fallback?

Testing that theory, it appears to hold true. By opening Firefox, adding a tab to a group, closing Firefox and reopening it, repeating these steps multiple times, I am able to confirm that in some cases the new tab is lost.

This makes me believe that there's some race condition between the following two functions associated with event listeners, or the cache, itself:

``` javascript
function onRemovedTab(tabId, { isWindowClosing, windowId }) {
    const log = logger.start('onRemovedTab', tabId, { isWindowClosing, windowId });

    if (isWindowClosing) {
>       reCreateTabsOnRemoveWindow.push(tabId);
        log.stop('add to reCreateTabsOnRemoveWindow');
    } else {
        Cache.removeTab(tabId);
        log.stop('tab removed from Cache');
    }
}
```


``` javascript
const onRemovedWindow = catchFunc(async function (windowId) {
    const log = logger.start(['info', 'onRemovedWindow'], windowId);

    let groupId = Cache.getWindowGroup(windowId);

    if (groupId) {
        sendMessage('window-closed', { windowId });
    }

    Cache.removeWindow(windowId);

>   let tabsToRestore = Cache.getTabsSessionAndRemove(reCreateTabsOnRemoveWindow);

    reCreateTabsOnRemoveWindow = [];

    if (tabsToRestore.length) {
        log.info('start merge tabs');
        let { tabsToRestore: prevRestore } = await Storage.get({ tabsToRestore: [] });
        tabsToRestore = tabsToRestore.filter(tab => !prevRestore.some(t => t.groupId === tab.groupId && t.url === tab.url && t.cookieStoreId === tab.cookieStoreId));
>       await Storage.set({
            tabsToRestore: [...prevRestore, ...tabsToRestore]
        });

        log.info('stop merge tabs > start restoring tabs');

        restoringMissedTabsPromise = tryRestoreMissedTabs();
        await restoringMissedTabsPromise.catch(log.onCatch('tryRestoreMissedTabs'));
        restoringMissedTabsPromise = null;
        log.info('stop restoring tabs');
    }

    log.stop();
});
```

**Comment by curtiola** [comment]
This happened to me yesterday after I enabled Firefox Sync. I was just trying to sync everything from windows to a fresh kubuntu install. I had a backup of my tab groups and my profile folder, which I needed to restore, disabling sync and using a STG backup wasn't enough. It's happening on both windows and kubuntu.

For some reason, it only happened to one profile, my 2nd profile synced just fine. I changed nothing other than enabling sync on both profiles and signing in to 2 fresh mozilla accounts I had just made. I have no idea why this one profile refuses to work, is there any useful information I can look for in settings or the profile folders?

**Comment by zkhcohen** [comment]
> Yet another edit: I think this is a firefox bug, not STG. It loses tabs that aren't in any groups as well, with or without "sync open tabs" enabled.

There may be another bug with Firefox sync, but this error occurs without it enabled.

I'm certain that the problem is due to a race condition, I just haven't had the time to fork the project and fix it.
